### ISSUE TITLE
SSR and Compile Time

### ISSUE THREAD

**Issue by bhough** [ricokahler/color2k#21]
First off, wanted to say that I love the creative approach here. Also, glad to hear that polished help inspire this work. Obviously the size difference has our interest piqued, but had some initial questions about the approach:

1. Being unfamiliar with the intricacies of how canvas interacts with the DOM, any limitations with this approach when using SSR rendering?
2. In polished, folks tend to either use our babel-plugin or others that preval functions where possible to remove the runtime impact and reduce bundle size. Given the canvas approach, I imagine there would need to be a custom plugin to render the canvas and do the work to precompile?

**Comment by ricokahler** [comment]
First: a quick update from what I posted in polished. Last night I switched the implementation from `canvas` to `getComputedStyle` because the [performance is much better](https://jsperf.com/polished-vs-canvas/3).

---

> Being unfamiliar with the intricacies of how canvas interacts with the DOM, any limitations with this approach when using SSR rendering?

I have a solve for this. If you look at the code, there is a hidden `require` in there that should only be run in non-browser environments.

https://github.com/ricokahler/color2k/blob/a48bf58729c8b2f60662ee22f75f1d786e31713e/packages/parse-to-rgba/src/index.ts#L27-L36

When that branch runs, it will parse the color using the optional node compatibility package <strike>`@color2k/node`</strike> `@color2k/compat` (which is a blatant copy-paste of polished 🙌) instead of going down the path of using the browser to parse colors.

> In polished, folks tend to either use our babel-plugin or others that preval functions where possible to remove the runtime impact and reduce bundle size. Given the canvas approach, I imagine there would need to be a custom plugin to render the canvas and do the work to precompile?

Given this approach, I think this will work fine in all node environments? I only tried this lib out in Gatsby and it worked flawlessly there but I still think I have more work to do regarding testing this thing.

Now that you remind me though, I forgot to test out the new `getComputedStyle` implementation in jsdom. I just added those tests and was greeted with this 😅:

![image](https://user-images.githubusercontent.com/10551026/81815622-35955200-94f8-11ea-9926-69ebd6fea6cc.png)

In summary, I think the shim + hidden require works pretty well but there's a decent amount of effort in testing this in non-node environments but at least [playwright is a big help](https://github.com/ricokahler/color2k/blob/a48bf58729c8b2f60662ee22f75f1d786e31713e/packages/parse-to-rgba/src/index.test.ts#L70-L83).

**Comment by ricokahler** [comment]
…and JSDOM is [now supported](https://github.com/ricokahler/color2k/releases/tag/v1.0.0-rc.4)!

**Comment by ricokahler** [comment]
I know this was posted as a question but I'd like to turn this issue into an actionable todo list in order to ensure this lib works in all non-browser environments (probably just testing locally). I don't expect any issues but it would be good to know before stamping the 1.0 on it.

---

To summarize (and to be referenced later in the README), in the browser, color2k will use [`getComputedStyle`](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle). In non-browser/SSR setups, an [inline `require`](https://github.com/ricokahler/color2k/blob/9b5875d8efcfe654cc69ca4ca4e02b9da7b5eb15/packages/parse-to-rgba/src/index.ts#L34) will pull the compat package from `node_modules`. Because the `require` is inline, this compat package will _not_ be included in the browser's bundle.

Here's a list environments/setups I've tested and what is left to test:

- [x] safari/webkit [(via playwright)](https://github.com/ricokahler/color2k/blob/9b5875d8efcfe654cc69ca4ca4e02b9da7b5eb15/packages/parse-to-rgba/src/index.test.ts#L156-L166) ✅
- [x] chrome/edge [(via playwright)](https://github.com/ricokahler/color2k/blob/9b5875d8efcfe654cc69ca4ca4e02b9da7b5eb15/packages/parse-to-rgba/src/index.test.ts#L156-L166) ✅
- [x] firefox [(via playwright)](https://github.com/ricokahler/color2k/blob/9b5875d8efcfe654cc69ca4ca4e02b9da7b5eb15/packages/parse-to-rgba/src/index.test.ts#L156-L166) ✅
- [ ] old edge
- [ ] IE 11
- [x] codesandbox ✅
- [x] create-react-app ✅ 
- [x] gatsby ✅ — requires `@color2k/compat` for SSR
- [x] node ✅ — requires `@color2k/compat`
- [x] jest + jsdom ✅ — requires `@color2k/compat`
- [ ] next.js  — will probably require `@color2k/compat` for SSR
- [ ] react-native — will probably require `@color2k/compat`
- [ ] preval — will probably require `@color2k/compat`
- [ ] vue/vue-cli
- [ ] angular/angular-cli
- [ ] svelte/svelte-cli
- [ ] deno

---

A final note on DX: The experience I'm going for is that `color2k` "just works" after an install. If it doesn't, then just install `@colork2/compat`. The intricacies of browser vs non-browser should be handled by the lib.
